webhooks/services.py

```diff
@@ -1,14 +1,137 @@
+            )
+        except JSONDecodeError:
+            # Don't raise an error here. It's probably more valuable to
+            # Return the data anyway than breaking the retrieval
+            return {"invalid JSON data": data}
+
+    def _to_json(self, data: List[bytes]) -> List[dict]:
+        data = [self._deserialise(val) for val in data]
+        try:
+            data = [json.loads(d) for d in data]
+        except JSONDecodeError:
+            # A multiline string has occured that has been split in
+            # an invalid way wrt JSON format
+
+            # Remove multiple spaces which occur when error msg is split into multiple lines
+            # turn off black formatter here because it replaces the regex
+            # fmt: off
+            data = [re.sub('\s+',' ', d) for d in data]
+            # Remove [..\" \"..] from the multiline error message because that is invalid json
+            data = [d.replace('\" \"', ' ') for d in data]
+            # fmt: on
+            data = [json.loads(d) for d in data]
+        return data
+
+    def _to_item(self, topic: str, data: List[dict]) -> List[dict]:
+        # Transform the data to the appropriate model
+        data = self._to_json(data=data)
+        if self._is_proof(topic):
+            data = [self._proof_hook_versioned(d) for d in data]
+        elif self._is_credential(topic):
+            data = [self._credential_hook_versioned(d) for d in data]
+        elif topic == "connections":
+            data = [self._version_connections(d) for d in data]
+        elif topic == "basicmessages":
+            data = [BasicMessagesHook(**d) for d in data]
+        return data
+
+    def _to_topic_item(self, topic: str, data: List[bytes]) -> List[TopicItem]:
+        data = self._to_item(topic=topic, data=data)
+        return (
+            [
+                TopicItem(
+                    topic=topic,
+                    wallet_id=payload.wallet_id,
+                    origin=payload.origin,
+                    payload=payload.dict(),
+                )
+                for payload in data
+            ]
+            if data
+            else []
+        )
+
+    async def add_topic_entry(self, topic: str, hook: bytes) -> List[TopicItem]:
         return await self._redis.sadd(topic, hook)
 
-    async def get_all_by_topic(self, topic: str) -> str:
-        return await self._redis.smembers(topic)
+    async def get_all_by_topic(self, topic: str) -> List[TopicItem]:
+        # Get the data from redis queue
+        data = await self._redis.smembers(topic)
+        # Transform it to TopicItems
+        items = self._to_topic_item(topic=topic, data=data)
+        for d in items:
+            del d.payload["wallet_id"]
+            del d.payload["origin"]
+        return items
+
+    async def get_all_for_topic_by_wallet_id(
+        self, topic: str, wallet_id: str
+    ) -> List[TopicItem]:
+        data = await self._redis.smembers(topic)
```

Comment:
Don't we store the items by both wallet id and topic? How is this going to scale? If not possible to index by both, it may be better to index by wallet_id as I suspect all by topic and the filtering for wallet id will be much more that filtering all by wallet id and then filtering for topic (I suspect there will be a looooooooot of wallets, that each are not that big)

Reply:
good shout

Reply:
Just to clarify - it was only stored by topic and wallet id was added to the payload. I was using redis sets so for a key such as connections I stored (added/appended all the hooks and added "wallet_id": "sthsth" to the actual hook. then I retrieved the hooks for the topic and filtered on the data. 

Either way it's a bit of a trade-off. If we really wanna scrimp on data in favour of speed/scalability here I reckon removing storing by topic and only storing by wallet_id is an option as by topic retrieval is really just an admin/debug kind of thing

Reply:
This is a bit of a tricky one. We either have to store the same data for different keys (same data for different keys is a [requested feature in redis](https://github.com/redis/redis/issues/2668)) or come up with some sort of trick.

Reply:
only using wallet id for now then